File: DoubleSquares/cpp/ds.cpp
// Double Squares solution in C++ for CodeEval.com by Steven A. Dunn

#include <cmath>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int getNumDoubleSquares(long number);

// length is the sqrt of the max input divided by 2
static int cachedVals[32767];

int main(int argc, char* argv[])
{
	ifstream inputFile(argv[1]);
	string line;
	
	if (inputFile)
	{
		getline(inputFile, line);
		while(getline(inputFile, line))
		{
			int result = getNumDoubleSquares(atol(line.c_str()));
			cout << result << endl;
		}
	}
}

int getNumDoubleSquares(long number)
{
	int count = 0;
	int upperLimit = sqrt(number/2);
	for (int i = 0; i <= upperLimit; i++)
	{
		if (cachedVals[i] == 0) { cachedVals[i] = i*i; }
		
		double j = sqrt((double)number - cachedVals[i]);
		if (j - (int)j == 0.0) { count++; }
	}
	return count; 
}

File: DoubleSquares/py3/ds.py
# Double Squares solution in Python 3 for CodeEval.com by Steven A. Dunn

import sys, math

cachedVals = [0] * 32767

f = open(sys.argv[1], 'r')
f.readline()

for line in f:
  N = int(line)
  count = 0
  upperLimit = math.sqrt(N/2)
  for i in range(0, int(upperLimit + 1)):
    if cachedVals[i] == 0: cachedVals[i] = i*i
    j = math.sqrt(N - cachedVals[i])
    if j - int(j) == 0.0: count += 1
  print (count)
